src/Perl6/Metamodel/Concretization.nqp

```diff
@@ -38,21 +35,29 @@ role Perl6::Metamodel::Concretization {
     }
 
     method !maybe_rebuild_table() {
-        my %new_conc_table := %!conc_table;
+        # Capturing the concretization list is first and foremost because we
+        # depend on its size to know whether or not a rebuild is necessary, but
+        # there may be a wait before then. Try for more predictable output.
+        my int $captured := nqp::elems(@!concretizations);
```

Comment:
Why isn't this inside the lock?

Reply:
It belonging in there would also bring the write in `add_concretization` into question. I erred on the side of the current behaviour there being correct. If a concretization gets added after a call to this has begun, then it shouldn't influence the resulting table. Taking the size of the list ASAP like this aims to prevent that from happening during the remainder of the call.